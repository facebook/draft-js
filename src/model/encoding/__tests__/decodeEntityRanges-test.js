/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 * @format
 */

'use strict';

jest.disableAutomock();

var decodeEntityRanges = require('decodeEntityRanges');
var {OrderedSet} = require('immutable');

var NONE = OrderedSet();
var SIX = OrderedSet.of('6');
var EIGHT = OrderedSet.of('8');

test('must decode when no entities present', () => {
  var decoded = decodeEntityRanges(' '.repeat(20), []);
  expect(decoded).toMatchSnapshot();
});

test('must decode when an entity is present', () => {
  var decoded = decodeEntityRanges(' '.repeat(5), [
    {
      offset: 2,
      length: 2,
      key: '6',
    },
  ]);
  expect(decoded).toMatchSnapshot();
});

test('must decode when multiple entities present', () => {
  var decoded = decodeEntityRanges(' '.repeat(8), [
    {
      offset: 2,
      length: 2,
      key: '6',
    },
    {
      offset: 5,
      length: 2,
      key: '8',
    },
  ]);
  expect(decoded).toMatchSnapshot();
});

test('must decode when an entity is present more than once', () => {
  var decoded = decodeEntityRanges(' '.repeat(8), [
    {
      offset: 2,
      length: 2,
      key: '6',
    },
    {
      offset: 5,
      length: 2,
      key: '6',
    },
  ]);
  expect(decoded).toMatchSnapshot();
});

test('must return overlapping ranges with multiple entities present', () => {
  var decoded = decodeEntityRanges(' '.repeat(8), [
    {
      offset: 2,
      length: 3,
      key: '6',
    },
    {
      offset: 4,
      length: 3,
      key: '8',
    },
  ]);
  expect(decoded).toMatchSnapshot();
});

it('must return interior ranges with multiple entities present', () => {
  var decoded = decodeEntityRanges(' '.repeat(8), [
    {
      offset: 2,
      length: 5,
      key: '6',
    },
    {
      offset: 4,
      length: 1,
      key: '8',
    },
  ]);
  expect(decoded).toMatchSnapshot();
});

test('must handle ranges that include surrogate pairs', () => {
  var decoded = decodeEntityRanges('Take a \uD83D\uDCF7 #selfie', [
    {
      offset: 6,
      length: 5,
      key: '6',
    },
    {
      offset: 13,
      length: 2,
      key: '8',
    },
  ]);
  expect(decoded).toMatchSnapshot();
});
