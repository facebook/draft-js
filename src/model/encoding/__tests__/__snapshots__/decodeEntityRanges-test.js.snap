// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must decode when an entity is present 1`] = `
Array [
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [],
]
`;

exports[`must decode when an entity is present more than once 1`] = `
Array [
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [],
]
`;

exports[`must decode when multiple entities present 1`] = `
Array [
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "8",
  ],
  Immutable.OrderedSet [
    "8",
  ],
  Immutable.OrderedSet [],
]
`;

exports[`must decode when no entities present 1`] = `
Array [
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
]
`;

exports[`must handle ranges that include surrogate pairs 1`] = `
Array [
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "8",
  ],
  Immutable.OrderedSet [
    "8",
  ],
  Immutable.OrderedSet [],
]
`;

exports[`must return interior ranges with multiple entities present 1`] = `
Array [
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
    "8",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [],
]
`;

exports[`must return overlapping ranges with multiple entities present 1`] = `
Array [
  Immutable.OrderedSet [],
  Immutable.OrderedSet [],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
  ],
  Immutable.OrderedSet [
    "6",
    "8",
  ],
  Immutable.OrderedSet [
    "8",
  ],
  Immutable.OrderedSet [
    "8",
  ],
  Immutable.OrderedSet [],
]
`;
